Comparison method violates its general contract.

```
com.starrocks.sql.common.DmlException: Refresh materialized view ads_play_basic_dev_educ_view_week failed after retrying 1 times(try-lock 0 times), error-msg : Comparison method violates its general contract!
	at com.starrocks.scheduler.PartitionBasedMvRefreshProcessor.doRefreshMaterializedViewWithRetry(PartitionBasedMvRefreshProcessor.java:395) ~[starrocks-fe.jar:?]
	at com.starrocks.scheduler.PartitionBasedMvRefreshProcessor.doMvRefresh(PartitionBasedMvRefreshProcessor.java:326) ~[starrocks-fe.jar:?]
	at com.starrocks.scheduler.PartitionBasedMvRefreshProcessor.processTaskRun(PartitionBasedMvRefreshProcessor.java:200) ~[starrocks-fe.jar:?]
	at com.starrocks.scheduler.TaskRun.executeTaskRun(TaskRun.java:257) ~[starrocks-fe.jar:?]
	at com.starrocks.scheduler.TaskRunExecutor.lambda$executeTaskRun$0(TaskRunExecutor.java:57) ~[starrocks-fe.jar:?]
	at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1700) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[?:?]
	at java.lang.Thread.run(Thread.java:829) ~[?:?]
Caused by: java.lang.IllegalArgumentException: Comparison method violates its general contract!
	at java.util.TimSort.mergeHi(TimSort.java:903) ~[?:?]
	at java.util.TimSort.mergeAt(TimSort.java:520) ~[?:?]
	at java.util.TimSort.mergeCollapse(TimSort.java:448) ~[?:?]
	at java.util.TimSort.sort(TimSort.java:245) ~[?:?]
	at java.util.Arrays.sort(Arrays.java:1515) ~[?:?]
	at java.util.stream.SortedOps$SizedRefSortingSink.end(SortedOps.java:353) ~[?:?]
	at java.util.stream.Sink$ChainedReference.end(Sink.java:258) ~[?:?]
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:485) ~[?:?]
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474) ~[?:?]
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913) ~[?:?]
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:?]
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578) ~[?:?]
	at com.starrocks.sql.common.PRangeCellPlus.toPRangeCellPlus(PRangeCellPlus.java:78) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.common.RangePartitionDiffer.generateBaseRefMap(RangePartitionDiffer.java:464) ~[starrocks-fe.jar:?]
	at com.starrocks.scheduler.mv.MVPCTRefreshRangePartitioner.syncAddOrDropPartitions(MVPCTRefreshRangePartitioner.java:126) ~[starrocks-fe.jar:?]
	at com.starrocks.scheduler.PartitionBasedMvRefreshProcessor.syncPartitions(PartitionBasedMvRefreshProcessor.java:876) ~[starrocks-fe.jar:?]
	at com.starrocks.scheduler.PartitionBasedMvRefreshProcessor.doRefreshMaterializedView(PartitionBasedMvRefreshProcessor.java:419) ~[starrocks-fe.jar:?]
	at com.starrocks.scheduler.PartitionBasedMvRefreshProcessor.doRefreshMaterializedViewWithRetry(PartitionBasedMvRefreshProcessor.java:367) ~[starrocks-fe.jar:?]
```

### Steps to reproduce the behavior (Required)

Write an unit tests like below:

```java
@Test
public void testComparison() throws AnalysisException {
    final String partitionName = "p1";
    final PartitionKey key1 = PartitionKey.ofDateTime(LocalDateTime.of(2025, 5, 27, 10, 0, 0));
    final PartitionKey key2 = PartitionKey.ofDateTime(LocalDateTime.of(2025, 5, 27, 11, 0, 0));
    final PartitionKey key3 = PartitionKey.ofDateTime(LocalDateTime.of(2025, 5, 27, 12, 0, 0));
    final PartitionKey key4 = PartitionKey.ofDateTime(LocalDateTime.of(2025, 5, 27, 13, 0, 0));

    Range<PartitionKey> r1 = Range.closed(key1, key2);
    PRangeCellPlus cell1 = new PRangeCellPlus(partitionName, r1);
    Assert.assertEquals(0, cell1.compareTo(cell1));

    Range<PartitionKey> r2 = Range.closed(key2, key3);
    PRangeCellPlus cell2 = new PRangeCellPlus(partitionName, r2);
    Assert.assertEquals(-1, cell1.compareTo(cell2));
    Assert.assertEquals(1, cell2.compareTo(cell1));

    Range<PartitionKey> r3 = Range.closed(key3, key4);
    PRangeCellPlus cell3 = new PRangeCellPlus(partitionName, r3);
    Assert.assertEquals(-1, cell2.compareTo(cell3));
    Assert.assertEquals(1, cell3.compareTo(cell2));
    Assert.assertEquals(-1, cell1.compareTo(cell3));
    Assert.assertEquals(1, cell3.compareTo(cell1));

    Range<PartitionKey> r4 = Range.closed(key1, key4);
    PRangeCellPlus cell4 = new PRangeCellPlus(partitionName, r4);
    PRangeCellPlus cell5 = new PRangeCellPlus(partitionName, r3);
    PRangeCellPlus cell6 = new PRangeCellPlus(partitionName, r1);
    Assert.assertEquals(0, cell4.compareTo(cell5));
    Assert.assertEquals(0, cell4.compareTo(cell6));
    Assert.assertEquals(0, cell5.compareTo(cell6));
}
```

It's easy to get a result like below:

```bash
[ERROR] testComparison  Time elapsed: 0.553 s  <<< FAILURE!
java.lang.AssertionError: expected:<0> but was:<1>
        at com.starrocks.sql.common.PRangeCellPlusTest.testComparison(PRangeCellPlusTest.java:56)

```

### Expected behavior (Required)

For comparison, we must ensure that when c1 is equal to c2 and c2 is equal to c3, c1 must be equal to c3.

### Real behavior (Required)

Currently, when c1 is equal to c2 and c2 is equal to c3, c1 might be not equal to c3.

### StarRocks version (Required)
 - 3.3.2

--- a/fe/fe-core/src/main/java/com/starrocks/catalog/MaterializedView.java
+++ b/fe/fe-core/src/main/java/com/starrocks/catalog/MaterializedView.java
@@ -104,7 +104,6 @@ import java.time.LocalDateTime;
 import java.time.ZoneId;
 import java.util.ArrayList;
 import java.util.Collection;
-import java.util.Collections;
 import java.util.HashMap;
 import java.util.LinkedHashMap;
 import java.util.List;
@@ -1702,21 +1701,31 @@ public class MaterializedView extends OlapTable implements GsonPreProcessable, G
      */
     private boolean isManyToManyPartitionRangeMapping(PRangeCell srcRange,
                                                       List<PRangeCell> dstRanges) {
-        int mid = Collections.binarySearch(dstRanges, srcRange);
-        if (mid < 0) {
+        if (dstRanges.isEmpty()) {
             return false;
         }
+        List<PartitionKey> lowerPoints = dstRanges.stream().map(
+                dstRange -> dstRange.getRange().lowerEndpoint()).toList();
+        List<PartitionKey> upperPoints = dstRanges.stream().map(
+                dstRange -> dstRange.getRange().upperEndpoint()).toList();
 
-        int lower = mid - 1;
-        if (lower >= 0 && dstRanges.get(lower).isIntersected(srcRange)) {
-            return true;
-        }
+        PartitionKey lower = srcRange.getRange().lowerEndpoint();
+        PartitionKey upper = srcRange.getRange().upperEndpoint();
 
-        int higher = mid + 1;
-        if (higher < dstRanges.size() && dstRanges.get(higher).isIntersected(srcRange)) {
-            return true;
+        // For an interval [l, r], if there exists another interval [li, ri] that intersects with it, this interval
+        // must satisfy l ≤ ri and r ≥ li. Therefore, if there exists a pos_a such that for all k < pos_a,
+        // ri[k] < l, and there exists a pos_b such that for all k > pos_b, li[k] > r, then all intervals between
+        // pos_a and pos_b might potentially intersect with the interval [l, r].
+        int posA = PartitionKey.findLastLessEqualInOrderedList(lower, upperPoints);
+        int posB = PartitionKey.findLastLessEqualInOrderedList(upper, lowerPoints);
+
+        int matched = 0;
+        for (int i = posA; i <= posB && matched <= 1; ++i) {
+            if (dstRanges.get(i).isIntersected(srcRange)) {
+                ++matched;
+            }
         }
-        return false;
+        return matched > 1;
     }
 
     private Map<Table, List<Column>> getBaseTablePartitionColumnMapImpl() {

--- a/fe/fe-core/src/main/java/com/starrocks/catalog/PartitionKey.java
+++ b/fe/fe-core/src/main/java/com/starrocks/catalog/PartitionKey.java
@@ -478,6 +478,21 @@ public class PartitionKey implements Comparable<PartitionKey>, Writable {
         return new PartitionKey(keyList, typeList);
     }
 
+    public static int findLastLessEqualInOrderedList(PartitionKey key, List<PartitionKey> list) {
+        int lower = 0;
+        int higher = list.size() - 1;
+        while (lower <= higher) {
+            int mid = (lower + higher) >> 1;
+            PartitionKey current = list.get(mid);
+            if (current.compareTo(key) <= 0) {
+                lower = mid + 1;
+            } else {
+                higher = mid - 1;
+            }
+        }
+        return Math.max(higher, 0);
+    }
+
     @Override
     public void write(DataOutput out) throws IOException {
         int count = keys.size();

--- a/fe/fe-core/src/main/java/com/starrocks/sql/common/PRangeCell.java
+++ b/fe/fe-core/src/main/java/com/starrocks/sql/common/PRangeCell.java
@@ -51,10 +51,11 @@ public final class PRangeCell extends PCell implements Comparable<PRangeCell> {
      */
     @Override
     public int compareTo(PRangeCell o) {
-        if (isIntersected(o)) {
-            return 0;
+        int res = this.range.lowerEndpoint().compareTo(o.range.lowerEndpoint());
+        if (res != 0) {
+            return res;
         }
-        return this.range.lowerEndpoint().compareTo(o.range.lowerEndpoint());
+        return this.range.upperEndpoint().compareTo(o.range.upperEndpoint());
     }
 
     /**

--- a/fe/fe-core/src/main/java/com/starrocks/sql/common/PRangeCellPlus.java
+++ b/fe/fe-core/src/main/java/com/starrocks/sql/common/PRangeCellPlus.java
@@ -18,6 +18,7 @@ import com.google.common.collect.Range;
 import com.starrocks.analysis.Expr;
 import com.starrocks.catalog.PartitionKey;
 
+import java.util.LinkedList;
 import java.util.List;
 import java.util.Map;
 import java.util.Objects;
@@ -51,6 +52,41 @@ public class PRangeCellPlus implements Comparable<PRangeCellPlus> {
 
     public boolean isIntersected(PRangeCellPlus o) {
         return cell.isIntersected(o.getCell());
+    }
+
+    /**
+     * Get intersected cells.
+     *
+     * @param cell PRangeCellPlus whose intersected cells will be found.
+     * @param ranges need to be sorted.
+     * @return intersected cells
+     */
+    public static List<PRangeCellPlus> getIntersectedCells(PRangeCellPlus cell, List<PRangeCellPlus> ranges) {
+        if (ranges.isEmpty()) {
+            return List.of();
+        }
+        List<PartitionKey> lowerPoints = ranges.stream().map(
+                dstRange -> dstRange.getCell().getRange().lowerEndpoint()).toList();
+        List<PartitionKey> upperPoints = ranges.stream().map(
+                dstRange -> dstRange.getCell().getRange().upperEndpoint()).toList();
+
+        PartitionKey lower = cell.getCell().getRange().lowerEndpoint();
+        PartitionKey upper = cell.getCell().getRange().upperEndpoint();
+
+        // For an interval [l, r], if there exists another interval [li, ri] that intersects with it, this interval
+        // must satisfy l ≤ ri and r ≥ li. Therefore, if there exists a pos_a such that for all k < pos_a,
+        // ri[k] < l, and there exists a pos_b such that for all k > pos_b, li[k] > r, then all intervals between
+        // pos_a and pos_b might potentially intersect with the interval [l, r].
+        int posA = PartitionKey.findLastLessEqualInOrderedList(lower, upperPoints);
+        int posB = PartitionKey.findLastLessEqualInOrderedList(upper, lowerPoints);
+
+        List<PRangeCellPlus> results = new LinkedList<>();
+        for (int i = posA; i <= posB; ++i) {
+            if (ranges.get(i).isIntersected(cell)) {
+                results.add(ranges.get(i));
+            }
+        }
+        return results;
     }
 
     public static List<PRangeCellPlus> toPRangeCellPlus(Map<String, PCell> rangeMap) {

--- a/fe/fe-core/src/main/java/com/starrocks/sql/common/RangePartitionDiffer.java
+++ b/fe/fe-core/src/main/java/com/starrocks/sql/common/RangePartitionDiffer.java
@@ -40,7 +40,6 @@ import org.threeten.extra.PeriodDuration;
 
 import java.time.LocalDate;
 import java.time.LocalDateTime;
-import java.util.Collections;
 import java.util.Comparator;
 import java.util.HashMap;
 import java.util.HashSet;
@@ -51,6 +50,7 @@ import java.util.Set;
 import java.util.function.Predicate;
 import java.util.stream.Collectors;
 
+import static com.starrocks.sql.common.PRangeCellPlus.getIntersectedCells;
 import static com.starrocks.sql.common.PRangeCellPlus.toPRangeCellPlus;
 
 // TODO: refactor all related code into this class
@@ -450,24 +450,15 @@ public final class RangePartitionDiffer extends PartitionDiffer {
             Preconditions.checkArgument(partitionExprs.size() == 1);
             List<PRangeCellPlus> baseRanges = toPRangeCellPlus(refreshedPartitionsMap, partitionExprs.get(0));
             for (PRangeCellPlus baseRange : baseRanges) {
-                int mid = Collections.binarySearch(mvRanges, baseRange);
-                if (mid < 0) {
+                List<PRangeCellPlus> intersectedRanges = getIntersectedCells(baseRange, mvRanges);
+                if (intersectedRanges.isEmpty()) {
                     initialBaseRefMap(result, baseTable, baseRange);
                     continue;
                 }
-                PRangeCellPlus mvRange = mvRanges.get(mid);
-                updateTableRefMap(result, baseTable, baseRange.getPartitionName(), mvRange.getPartitionName());
 
-                int lower = mid - 1;
-                while (lower >= 0 && mvRanges.get(lower).isIntersected(baseRange)) {
-                    updateTableRefMap(result, baseTable, baseRange.getPartitionName(), mvRanges.get(lower).getPartitionName());
-                    lower--;
-                }
-
-                int higher = mid + 1;
-                while (higher < mvRanges.size() && mvRanges.get(higher).isIntersected(baseRange)) {
-                    updateTableRefMap(result, baseTable, baseRange.getPartitionName(), mvRanges.get(higher).getPartitionName());
-                    higher++;
+                for (PRangeCellPlus intersectedRange : intersectedRanges) {
+                    updateTableRefMap(
+                            result, baseTable, baseRange.getPartitionName(), intersectedRange.getPartitionName());
                 }
             }
         }
@@ -505,25 +496,14 @@ public final class RangePartitionDiffer extends PartitionDiffer {
             initialMvRefMap(result, mvRanges, entry.getKey());
             List<PRangeCellPlus> baseRanges = entry.getValue();
             for (PRangeCellPlus baseRange : baseRanges) {
-                int mid = Collections.binarySearch(mvRanges, baseRange);
-                if (mid < 0) {
+                List<PRangeCellPlus> intersectedRanges = getIntersectedCells(baseRange, mvRanges);
+                if (intersectedRanges.isEmpty()) {
                     continue;
                 }
-                PRangeCellPlus mvRange = mvRanges.get(mid);
-                updatePartitionRefMap(result, mvRange.getPartitionName(), entry.getKey(), baseRange.getPartitionName());
 
-                int lower = mid - 1;
-                while (lower >= 0 && mvRanges.get(lower).isIntersected(baseRange)) {
+                for (PRangeCellPlus intersectedRange : intersectedRanges) {
                     updatePartitionRefMap(
-                            result, mvRanges.get(lower).getPartitionName(), entry.getKey(), baseRange.getPartitionName());
-                    lower--;
-                }
-
-                int higher = mid + 1;
-                while (higher < mvRanges.size() && mvRanges.get(higher).isIntersected(baseRange)) {
-                    updatePartitionRefMap(
-                            result, mvRanges.get(higher).getPartitionName(), entry.getKey(), baseRange.getPartitionName());
-                    higher++;
+                            result, intersectedRange.getPartitionName(), entry.getKey(), baseRange.getPartitionName());
                 }
             }
         }

--- a/fe/fe-core/src/main/java/com/starrocks/sql/common/SyncPartitionUtils.java
+++ b/fe/fe-core/src/main/java/com/starrocks/sql/common/SyncPartitionUtils.java
@@ -327,25 +327,26 @@ public class SyncPartitionUtils {
 
         Collections.sort(srcRanges, PRangeCellPlus::compareTo);
         Collections.sort(dstRanges, PRangeCellPlus::compareTo);
-
+        List<PartitionKey> lowerPoints = dstRanges.stream().map(
+                dstRange -> dstRange.getCell().getRange().lowerEndpoint()).toList();
+        List<PartitionKey> upperPoints = dstRanges.stream().map(
+                dstRange -> dstRange.getCell().getRange().upperEndpoint()).toList();
         for (PRangeCellPlus srcRange : srcRanges) {
-            int mid = Collections.binarySearch(dstRanges, srcRange);
-            if (mid < 0) {
-                continue;
-            }
+            PartitionKey lower = srcRange.getCell().getRange().lowerEndpoint();
+            PartitionKey upper = srcRange.getCell().getRange().upperEndpoint();
+
+            // For an interval [l, r], if there exists another interval [li, ri] that intersects with it, this interval
+            // must satisfy l ≤ ri and r ≥ li. Therefore, if there exists a pos_a such that for all k < pos_a,
+            // ri[k] < l, and there exists a pos_b such that for all k > pos_b, li[k] > r, then all intervals between
+            // pos_a and pos_b might potentially intersect with the interval [l, r].
+            int posA = PartitionKey.findLastLessEqualInOrderedList(lower, upperPoints);
+            int posB = PartitionKey.findLastLessEqualInOrderedList(upper, lowerPoints);
+
             Set<String> addedSet = result.get(srcRange.getPartitionName());
-            addedSet.add(dstRanges.get(mid).getPartitionName());
-
-            int lower = mid - 1;
-            while (lower >= 0 && dstRanges.get(lower).isIntersected(srcRange)) {
-                addedSet.add(dstRanges.get(lower).getPartitionName());
-                lower--;
-            }
-
-            int higher = mid + 1;
-            while (higher < dstRanges.size() && dstRanges.get(higher).isIntersected(srcRange)) {
-                addedSet.add(dstRanges.get(higher).getPartitionName());
-                higher++;
+            for (int i = posA; i <= posB; ++i) {
+                if (dstRanges.get(i).isIntersected(srcRange)) {
+                    addedSet.add(dstRanges.get(i).getPartitionName());
+                }
             }
         }
         return result;

--- a/fe/fe-core/src/test/java/com/starrocks/catalog/PartitionKeyTest.java
+++ b/fe/fe-core/src/test/java/com/starrocks/catalog/PartitionKeyTest.java
@@ -36,6 +36,7 @@ package com.starrocks.catalog;
 
 import com.starrocks.common.AnalysisException;
 import com.starrocks.common.FeConstants;
+import com.starrocks.common.util.DateUtils;
 import com.starrocks.server.GlobalStateMgr;
 import com.starrocks.sql.ast.PartitionValue;
 import org.junit.Assert;
@@ -215,5 +216,25 @@ public class PartitionKeyTest {
         // 3. delete files
         dis.close();
         file.delete();
+    }
+
+    private PartitionKey createPartitionKey(String key) throws AnalysisException {
+        return PartitionKey.ofDateTime(DateUtils.parseStrictDateTime(key));
+    }
+
+    @Test
+    public void testPartitionKeys() throws AnalysisException {
+        List<PartitionKey> list = Arrays.asList(
+                createPartitionKey("20230801"),
+                createPartitionKey("20230802"),
+                createPartitionKey("20230804"),
+                createPartitionKey("20230805"),
+                createPartitionKey("20230806"),
+                createPartitionKey("20230807"));
+        Assert.assertEquals(0, PartitionKey.findLastLessEqualInOrderedList(createPartitionKey("20230731"), list));
+        Assert.assertEquals(0, PartitionKey.findLastLessEqualInOrderedList(createPartitionKey("20230801"), list));
+        Assert.assertEquals(1, PartitionKey.findLastLessEqualInOrderedList(createPartitionKey("20230802"), list));
+        Assert.assertEquals(1, PartitionKey.findLastLessEqualInOrderedList(createPartitionKey("20230803"), list));
+        Assert.assertEquals(5, PartitionKey.findLastLessEqualInOrderedList(createPartitionKey("20230808"), list));
     }
 }

--- a/fe/fe-core/src/test/java/com/starrocks/sql/common/PRangeCellPlusTest.java
+++ b/fe/fe-core/src/test/java/com/starrocks/sql/common/PRangeCellPlusTest.java
@@ -1,0 +1,58 @@
+// Copyright 2021-present StarRocks, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.starrocks.sql.common;
+
+import com.google.common.collect.Range;
+import com.starrocks.catalog.PartitionKey;
+import com.starrocks.common.AnalysisException;
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.time.LocalDateTime;
+
+public class PRangeCellPlusTest {
+    @Test
+    public void testComparison() throws AnalysisException {
+        final String partitionName = "p1";
+        final PartitionKey key1 = PartitionKey.ofDateTime(LocalDateTime.of(2025, 5, 27, 10, 0, 0));
+        final PartitionKey key2 = PartitionKey.ofDateTime(LocalDateTime.of(2025, 5, 27, 11, 0, 0));
+        final PartitionKey key3 = PartitionKey.ofDateTime(LocalDateTime.of(2025, 5, 27, 12, 0, 0));
+        final PartitionKey key4 = PartitionKey.ofDateTime(LocalDateTime.of(2025, 5, 27, 13, 0, 0));
+
+        Range<PartitionKey> r1 = Range.closed(key1, key2);
+        PRangeCellPlus cell1 = new PRangeCellPlus(partitionName, r1);
+        Assert.assertEquals(0, cell1.compareTo(cell1));
+
+        Range<PartitionKey> r2 = Range.closed(key2, key3);
+        PRangeCellPlus cell2 = new PRangeCellPlus(partitionName, r2);
+        Assert.assertEquals(-1, cell1.compareTo(cell2));
+        Assert.assertEquals(1, cell2.compareTo(cell1));
+
+        Range<PartitionKey> r3 = Range.closed(key3, key4);
+        PRangeCellPlus cell3 = new PRangeCellPlus(partitionName, r3);
+        Assert.assertEquals(-1, cell2.compareTo(cell3));
+        Assert.assertEquals(1, cell3.compareTo(cell2));
+        Assert.assertEquals(-1, cell1.compareTo(cell3));
+        Assert.assertEquals(1, cell3.compareTo(cell1));
+
+        Range<PartitionKey> r4 = Range.closed(key1, key4);
+        PRangeCellPlus cell4 = new PRangeCellPlus(partitionName, r4);
+        PRangeCellPlus cell5 = new PRangeCellPlus(partitionName, r3);
+        PRangeCellPlus cell6 = new PRangeCellPlus(partitionName, r1);
+        Assert.assertEquals(-1, cell4.compareTo(cell5));
+        Assert.assertEquals(1, cell4.compareTo(cell6));
+        Assert.assertEquals(1, cell5.compareTo(cell6));
+    }
+}

--- a/fe/fe-core/src/test/java/com/starrocks/sql/common/PRangeCellTest.java
+++ b/fe/fe-core/src/test/java/com/starrocks/sql/common/PRangeCellTest.java
@@ -1,0 +1,85 @@
+// Copyright 2021-present StarRocks, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.starrocks.sql.common;
+
+import com.google.common.collect.Range;
+import com.starrocks.catalog.PartitionKey;
+import com.starrocks.common.AnalysisException;
+import com.starrocks.common.util.DateUtils;
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.time.LocalDateTime;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+public class PRangeCellTest {
+
+    private PartitionKey createPartitionKey(String key) throws AnalysisException {
+        return PartitionKey.ofDateTime(DateUtils.parseStrictDateTime(key));
+    }
+
+    private PRangeCell buildPartitionRange(String start, String end) throws AnalysisException {
+        return new PRangeCell(Range.closedOpen(createPartitionKey(start), createPartitionKey(end)));
+    }
+
+    @Test
+    public void testComparison() throws AnalysisException {
+        final PartitionKey key1 = PartitionKey.ofDateTime(LocalDateTime.of(2025, 5, 27, 10, 0, 0));
+        final PartitionKey key2 = PartitionKey.ofDateTime(LocalDateTime.of(2025, 5, 27, 11, 0, 0));
+        final PartitionKey key3 = PartitionKey.ofDateTime(LocalDateTime.of(2025, 5, 27, 12, 0, 0));
+        final PartitionKey key4 = PartitionKey.ofDateTime(LocalDateTime.of(2025, 5, 27, 13, 0, 0));
+
+        Range<PartitionKey> r1 = Range.closed(key1, key2);
+        PRangeCell cell1 = new PRangeCell(r1);
+        Assert.assertEquals(0, cell1.compareTo(cell1));
+
+        Range<PartitionKey> r2 = Range.closed(key2, key3);
+        PRangeCell cell2 = new PRangeCell(r2);
+        Assert.assertEquals(-1, cell1.compareTo(cell2));
+        Assert.assertEquals(1, cell2.compareTo(cell1));
+
+        Range<PartitionKey> r3 = Range.closed(key3, key4);
+        PRangeCell cell3 = new PRangeCell(r3);
+        Assert.assertEquals(-1, cell2.compareTo(cell3));
+        Assert.assertEquals(1, cell3.compareTo(cell2));
+        Assert.assertEquals(-1, cell1.compareTo(cell3));
+        Assert.assertEquals(1, cell3.compareTo(cell1));
+
+        Range<PartitionKey> r4 = Range.closed(key1, key4);
+        PRangeCell cell4 = new PRangeCell(r4);
+        PRangeCell cell5 = new PRangeCell(r3);
+        PRangeCell cell6 = new PRangeCell(r1);
+        Assert.assertEquals(-1, cell4.compareTo(cell5));
+        Assert.assertEquals(1, cell4.compareTo(cell6));
+        Assert.assertEquals(1, cell5.compareTo(cell6));
+    }
+
+    @Test
+    public void testBinarySearch() throws AnalysisException {
+        List<PRangeCell> srcs = Arrays.asList(
+                buildPartitionRange("00000101", "20230801"),
+                buildPartitionRange("20230801", "20230802"),
+                buildPartitionRange("20230802", "20230803"),
+                buildPartitionRange("20230803", "20230804"),
+                buildPartitionRange("20230804", "20230805"),
+                buildPartitionRange("20230805", "20991231"));
+
+        Collections.sort(srcs, PRangeCell::compareTo);
+
+        Assert.assertEquals(1, Collections.binarySearch(srcs, buildPartitionRange("20230801", "20230802")));
+    }
+}

--- a/fe/fe-core/src/test/java/com/starrocks/sql/common/SyncPartitionBench.java
+++ b/fe/fe-core/src/test/java/com/starrocks/sql/common/SyncPartitionBench.java
@@ -311,24 +311,27 @@ public class SyncPartitionBench {
         Collections.sort(srcRanges, PRangeCellPlus::compareTo);
         Collections.sort(dstRanges, PRangeCellPlus::compareTo);
 
+        List<PartitionKey> lowerPoints = dstRanges.stream().map(
+                dstRange -> dstRange.getCell().getRange().lowerEndpoint()).toList();
+        List<PartitionKey> upperPoints = dstRanges.stream().map(
+                dstRange -> dstRange.getCell().getRange().upperEndpoint()).toList();
+
         for (PRangeCellPlus srcRange : srcRanges) {
-            int mid = Collections.binarySearch(dstRanges, srcRange);
-            if (mid < 0) {
-                continue;
-            }
+            PartitionKey lower = srcRange.getCell().getRange().lowerEndpoint();
+            PartitionKey upper = srcRange.getCell().getRange().upperEndpoint();
+
+            // For an interval [l, r], if there exists another interval [li, ri] that intersects with it, this interval
+            // must satisfy l ≤ ri and r ≥ li. Therefore, if there exists a pos_a such that for all k < pos_a,
+            // ri[k] < l, and there exists a pos_b such that for all k > pos_b, li[k] > r, then all intervals between
+            // pos_a and pos_b might potentially intersect with the interval [l, r].
+            int posA = PartitionKey.findLastLessEqualInOrderedList(lower, upperPoints);
+            int posB = PartitionKey.findLastLessEqualInOrderedList(upper, lowerPoints);
+
             Set<String> addedSet = result.get(srcRange.getPartitionName());
-            addedSet.add(dstRanges.get(mid).getPartitionName());
-
-            int lower = mid - 1;
-            while (lower >= 0 && dstRanges.get(lower).isIntersected(srcRange)) {
-                addedSet.add(dstRanges.get(lower).getPartitionName());
-                lower--;
-            }
-
-            int higher = mid + 1;
-            while (higher < dstRanges.size() && dstRanges.get(higher).isIntersected(srcRange)) {
-                addedSet.add(dstRanges.get(higher).getPartitionName());
-                higher++;
+            for (int i = posA; i <= posB; ++i) {
+                if (dstRanges.get(i).isIntersected(srcRange)) {
+                    addedSet.add(dstRanges.get(i).getPartitionName());
+                }
             }
         }
         return result;

